src/modules/roc_core/target_libunwind/roc_core/backtrace.cpp

```diff
@@ -0,0 +1,189 @@
+	} while(number > 0);
+
+	while(++i < MaxDigits+1)
+		output_string[j++] = output_string[i];
+	output_string[j] = 0;
+}
+
+/* Checking if backtrace is empty or not.
+ * This must be signal safe.
+ */
+bool is_backtrace_available()
+{
+	/* To point to the current frame in the call stack.
+	 */
+	unw_cursor_t cursor;
+
+	/* To store the snapshot of the CPU registers.
+	 */
+	unw_context_t context;
+
+	/* To get snapshot of the CPU register.
+	 * unw_getcontext() is signal safe.
+	 */
+	if(unw_getcontext(&context) < 0)
+		return false;
+
+	/* To point to current frame.
+	 * unw_init_local() is signal safe.
+	 */
+	if(unw_init_local(&cursor, &context) < 0)
+		return false;
+
+	/* Moving to previously called frames & going through each of them.
+	 * unw_step() is signal safe.
+	 * If there is atleast one entry in backtrace
+	 */
+	if (unw_step(&cursor) > 0)
+		return true;
+	return false;
+}
+
+/* Print function name, offset, instruction pointer address
+ * This must be signal safe.
+ */
+void backtrace_symbols_fd(int fd){
+	/* To point to the current frame in the call stack.
+	 */
+	unw_cursor_t cursor;
+
+	/* To store the snapshot of the CPU registers.
+	 */
+	unw_context_t context;
+
+	/* To get snapshot of the CPU register.
+	 * unw_getcontext() is signal safe.
+	 */
+	unw_getcontext(&context);
+
+	/* To point to current frame.
+	 * unw_init_local() is signal safe. 
+	 */
+	unw_init_local(&cursor, &context);
+
+	/* Moving to previously called frames & going through each of them.
+	 * unw_step() is signal safe.
+	 * buffer = final output that will be written to stderr. 
+	 */
+	uint32_t index = 0;
+	while (unw_step(&cursor) > 0) {
+		char buffer[DefaultLen] = "#";
+		index++;
+		char function_name[128];
+		unw_word_t offset, ip;
+		function_name[0] = '\0';
+		/* Get value stored in instruction register
+		 * unw_get_reg() is signal safe.
+		 */
+		unw_get_reg(&cursor, UNW_REG_IP, &ip);
+		/* Get function name & offset
+		 * unw_get_proc_name() is signal safe.
+		 * 'status' checks if unw_get_proc_name() is successful or not.
+		 * 'status = 0', successfully executed the function.
+		 * if status is negative, then we'll stop.
+		 * This is optional and might be useful for debugging.
+		 */
+		int32_t status = unw_get_proc_name(&cursor, function_name, 40, &offset);
+		if(status < 0) {
+			if(status == UNW_EUNSPEC) {
+				write(fd, "An unspecified error occured.", 30);
+			}
+			else if(status == UNW_ENOINFO) {
+				write(fd, "Unable to determine name of procedure.", 39);
+			}
+			else if(status == UNW_ENOMEM) {
+				write(fd, "Procedure was too long to fit in the buffer.", 50);
+			}
+			else {
+				write(fd, "Error not found.", 16);
+			}
+		}
+		else {
+			/* Printing => index, function name, offset, ip
+			 * using write(). 
+			 * write() & strcat() are signal safe.
+			 */
+			char index_buffer[MaxDigits+1];
+			uint32_to_string(index, MaxDigits, index_buffer);
+			strcat(buffer, index_buffer);
+			strcat(buffer, ": (");
```

Comment:
We still need to check buffer boundaries in all these strcat calls.

Reply:
I will just check this once more.

Reply:
By checking buffer, do you mean say that 200 is not enough for `buffer` array size? or do you want me to do some kind of check before calling strcat()?

Reply:
The second. Since the line length is not known for sure, we should implement boundary checking.

For example, you can implement a helper function like:

```
void safe_strcat(char* buffer, size_t buffer_size, const char* str) {
    ...
}
```

... which will truncate the string if it is too long and ensure that we never go outside the buffer boundaries, and then use it instead of strcat().

Reply:
Can I use `strncat()` in that case? http://man7.org/linux/man-pages/man3/strncat.3.html I can specify how many characters can be copied to the buffer.

Reply:
Take a look at the strncat() limits the size of the `src` but it does not take into account the size of `dest` so it has the same problem.

You could use strlcat() for that but it's non-portable so we can't use it in target_libunwind.